contracts/cosmwasm-vm/cw-cluster-connection/src/contract.rs

```diff
@@ -0,0 +1,302 @@
+use cw_xcall_lib::network_address::NetId;
+
+use super::*;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cluster-connection";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const DEFAULT_SIGNATURE_THRESHOLD: u8 = 1;
+
+impl<'a> ClusterConnection<'a> {
+    pub fn instantiate(
+        &mut self,
+        deps: DepsMut,
+        _env: Env,
+        _info: MessageInfo,
+        msg: InstantiateMsg,
+    ) -> Result<Response, ContractError> {
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        let xcall_address = deps.api.addr_validate(&msg.xcall_address)?;
+        self.store_xcall(deps.storage, xcall_address)?;
+
+        self.store_admin(deps.storage, _info.sender)?;
+
+        let relayer = deps.api.addr_validate(&msg.relayer)?;
+        self.store_relayer(deps.storage, relayer)?;
+
+        self.store_denom(deps.storage, msg.denom)?;
+
+        let _ = self.store_conn_sn(deps.storage, 0);
+
+        self.store_signature_threshold(deps.storage, DEFAULT_SIGNATURE_THRESHOLD)?;
+
+        Ok(Response::new()
+            .add_attribute("action", "instantiate")
+            .add_attribute("relayer", msg.relayer)
+            .add_attribute("xcall_address", msg.xcall_address))
+    }
+
+    pub fn set_admin(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        address: Addr,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+        let new_admin = deps.api.addr_validate(address.as_str())?;
+        let _ = self.store_admin(deps.storage, new_admin);
+        Ok(Response::new().add_attribute("action", "set_admin"))
+    }
+
+    pub fn set_relayer(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        address: Addr,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+        let new_relayer = deps.api.addr_validate(address.as_str())?;
+        let _ = self.store_relayer(deps.storage, new_relayer);
+        Ok(Response::new().add_attribute("action", "set_relayer"))
+    }
+
+    pub fn set_validators(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        validators: Vec<String>,
+        threshold: u8,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+
+        if threshold < 1 {
+            return Err(ContractError::InvalidThreshold {
+                msg: "threshold should be at least 1".to_string(),
+            });
+        }
+
+        if validators.len() > 0 {
+            self.clear_validators(deps.storage)?;
+            for rlr in validators {
+                self.store_validator(deps.storage, rlr)?;
+            }
+        }
+
+        let validators_set = self.get_validators(deps.storage)?;
+
+        if threshold as usize > validators_set.len() {
+            return Err(ContractError::InvalidThreshold {
```

Comment:
length of validators should be greater than threshold. if threshold is greater than validator len it will never be approved.